README.md
# Underbelly Buddy

Available to download on Curse by clicking [here](https://mods.curse.com/addons/wow/257332-underbelly-buddy) or via the [Curse client](https://www.curse.com/). You can also check out this project on [CurseForge](https://wow.curseforge.com/projects/underbelly-buddy).

### About

A World of Wacraft addon that tracks the time of the hired guard buff obtainable in the Underbelly of Dalaran. 
The hired guard prevents you from PvP flagging when the Underbelly turns into a PvP free for all.

### Info

- The timer bar will not persist through UI reloads.
- The timer bar can be dismissed by right clicking on it.

### Options

Found under **Interface > AddOns Tab > UnderbellyBuddy** or using the slash command `/ub`

### Slash Commands

- `/ub` - Brings up the options menu
- `/ubshow` - Shows the bar if you dismissed it away by right clicking it
- `/ubhide` - Hides the bar if no longer wish to see it

### Found a Bug?

Please open an issue on Curse: [https://wow.curseforge.com/projects/underbelly-buddy/issues](https://wow.curseforge.com/projects/underbelly-buddy/issues)


CHANGELOG.md
# Change Log

## 1.0.4 (03/30/2017)

**Imporvements:**

- Updated version for 7.2
- Minor code cleanup.

## 1.0.3 (01/30/2017)

**Imporvements:**

- Added warning sound when warning message pops up.

## 1.0.2 (01/30/2017)

**Fixed bugs:**

- Fixed issue with the not including the correct libs.

**Imporvements:**

- Using `/ub` will now bring up the same options menu found in Blizzard's options.
- New commands `/ubshow` and `/ubhide` will now show and hide the timer bar.

## 1.0.1 (01/19/2017)

**Fixed bugs:**

- Fixed issues with hiding the timer bar.

## 1.0.0 (01/16/2017)

**Features:**

- Tracks the time of the hired guard buff obtainable in the Underbelly of Dalaran.


Modules/Config.lua
ub = LibStub('AceAddon-3.0'):GetAddon('UnderbellyBuddy')
ubHiredGuard = ub:GetModule('HiredGuard')
ubConfig = ub:NewModule('UnderbellyBuddyConfig', 'AceConsole-3.0')

local AceConfig = LibStub('AceConfig-3.0')
local AceConfigDialog = LibStub('AceConfigDialog-3.0')

function ubConfig:OnInitialize()
    self.db = ub.db

    local options = self:GetOptions()
    
    options.args.profile = LibStub('AceDBOptions-3.0'):GetOptionsTable(self.db)

    AceConfig:RegisterOptionsTable('UnderbellyBuddy', options)
    AceConfigDialog:AddToBlizOptions('UnderbellyBuddy', 'UnderbellyBuddy', nil, 'main') 
    AceConfigDialog:AddToBlizOptions('UnderbellyBuddy', 'Profiles', 'UnderbellyBuddy', 'profile')

    self:RegisterChatCommand('ub', 'OpenOptions')

    self.db.RegisterCallback(self, 'OnProfileChanged', 'RefreshConfig')
    self.db.RegisterCallback(self, 'OnProfileCopied', 'RefreshConfig')
    self.db.RegisterCallback(self, 'OnProfileReset', 'RefreshConfig')
end

function ubConfig:OpenOptions()
    AceConfigDialog:SetDefaultSize('UnderbellyBuddy', 600, 425)
    AceConfigDialog:Open('UnderbellyBuddy')
end

function ubConfig:RefreshConfig()
    ubHiredGuard:SetSize(_, self.db.profile.size)
end

function ubConfig:GetOptions()
    return {
        name = 'UnderbellyBuddy',
        handler = ub,
        type = 'group',
        args = {
            main = {
                order = 1,
                name = 'Main Options',
                type = 'group',
                args = {
                    header1 = {
                    order = 0,
                        name = 'Options',
                        type = 'header'
                    },
                    enable = {
                        order = 0.1,
                        name = 'Enable',
                        desc = 'Enables / disables the addon',
                        type = 'toggle',
                        width = 'full',
                        set = function(_, value) ubHiredGuard:SetEnabled(_, value) end,
                        get = function() return ub.db.profile.enable end
                    },
                    lock = {
                        order = 0.2,
                        name = 'Lock Timer',
                        desc = 'Locks the timer bar in place',
                        width = 'full',
                        type = 'toggle',
                        set = function(_, value) ub.db.profile.lock = value end,
                        get = function() return ub.db.profile.lock end
                    },
                    warnings = {
                        order = 0.3,
                        name = 'Display Warnings',
                        desc = 'Displays warning messages after a certain amount of time',
                        width = 'full',
                        type = 'toggle',
                        set = function(_, value) ub.db.profile.warning = value end,
                        get = function() return ub.db.profile.warning end
                    },
                    header2 = {
                    order = 1,
                        name = 'Appearance',
                        type = 'header'
                    },
                    bar = {
                        order = 1.1,
                        name = 'Test Bar',
                        desc = 'Shows a test bar to move or adjust size',
                        width = 'full',
                        type = 'execute',
                        func = function() ubHiredGuard:ShowTestBar() end
                    },
                    size = {
                        order = 1.2,
                        name = 'Bar Size',
                        desc = 'Changes the size of the timer bar',
                        width = 'full',
                        type = 'range',
                        min = 1,
                        max = 5,
                        set =  function(_, value) ubHiredGuard:SetSize(_, value) end,
                        get = function() return ub.db.profile.size end
                    },
                    header3 = {
                        order = 2,
                        name = 'About',
                        type = 'header'
                    },
                    about = {
                        order = 2.1,
                        name = 'Version: @project-version@ Created by Pigletoos of Skywall',
                        type = 'description'
                    },
                }
            }
        }
    }
end

Modules/Options.lua
local UnderbellyBuddy = LibStub('AceAddon-3.0'):GetAddon('UnderbellyBuddy')
local UnderbellyBuddyHiredGuard = UnderbellyBuddy:GetModule('UnderbellyBuddyHiredGuard')
local UnderbellyBuddyOptions = UnderbellyBuddy:NewModule('UnderbellyBuddyOptions', 'AceConsole-3.0')
local AceConfig = LibStub('AceConfig-3.0')
local AceConfigDialog = LibStub('AceConfigDialog-3.0')

function UnderbellyBuddyOptions:OnInitialize()
    self.db = UnderbellyBuddy.db
    
    local options = self:GetOptions()
    
    options.args.profile = LibStub('AceDBOptions-3.0'):GetOptionsTable(self.db)
    
    AceConfig:RegisterOptionsTable('UnderbellyBuddy', options)
    AceConfigDialog:AddToBlizOptions('UnderbellyBuddy', 'UnderbellyBuddy', nil, 'main')
    AceConfigDialog:AddToBlizOptions('UnderbellyBuddy', 'Profiles', 'UnderbellyBuddy', 'profile')
    
    self:RegisterChatCommand('ub', 'OpenOptions')
    
    self.db.RegisterCallback(self, 'OnProfileChanged', 'RefreshConfig')
    self.db.RegisterCallback(self, 'OnProfileCopied', 'RefreshConfig')
    self.db.RegisterCallback(self, 'OnProfileReset', 'RefreshConfig')
end

function UnderbellyBuddyOptions:OpenOptions()
    AceConfigDialog:SetDefaultSize('UnderbellyBuddy', 600, 425)
    AceConfigDialog:Open('UnderbellyBuddy')
end

function UnderbellyBuddyOptions:RefreshConfig()
    UnderbellyBuddyHiredGuard:SetSize(_, self.db.profile.size)
end

function UnderbellyBuddyOptions:GetOptions()
    return {
        name = 'UnderbellyBuddy',
        handler = UnderbellyBuddy,
        type = 'group',
        args = {
            main = {
                order = 1,
                name = 'Main Options',
                type = 'group',
                args = {
                    header1 = {
                        order = 0,
                        name = 'Options',
                        type = 'header'
                    },
                    enable = {
                        order = 0.1,
                        name = 'Enable',
                        desc = 'Enables / disables the addon',
                        type = 'toggle',
                        width = 'full',
                        set = function(_, value) UnderbellyBuddyHiredGuard:SetEnabled(_, value) end,
                        get = function() return self.db.profile.enable end
                    },
                    lock = {
                        order = 0.2,
                        name = 'Lock Timer',
                        desc = 'Locks the timer bar in place',
                        width = 'full',
                        type = 'toggle',
                        set = function(_, value) self.db.profile.lock = value end,
                        get = function() return self.db.profile.lock end
                    },
                    warnings = {
                        order = 0.3,
                        name = 'Display Warnings',
                        desc = 'Displays warning messages after a certain amount of time',
                        width = 'full',
                        type = 'toggle',
                        set = function(_, value) self.db.profile.warning = value end,
                        get = function() return  self.db.profile.warning end
                    },
                    header2 = {
                        order = 1,
                        name = 'Appearance',
                        type = 'header'
                    },
                    bar = {
                        order = 1.1,
                        name = 'Test Bar',
                        desc = 'Shows a test bar to move or adjust size',
                        width = 'full',
                        type = 'execute',
                        func = function() UnderbellyBuddyHiredGuard:ShowTestBar() end
                    },
                    size = {
                        order = 1.2,
                        name = 'Bar Size',
                        desc = 'Changes the size of the timer bar',
                        width = 'full',
                        type = 'range',
                        min = 1,
                        max = 5,
                        set = function(_, value) UnderbellyBuddyHiredGuard:SetSize(_, value) end,
                        get = function() return self.db.profile.size end
                    },
                    header3 = {
                        order = 2,
                        name = 'About',
                        type = 'header'
                    },
                    about = {
                        order = 2.1,
                        name = 'Version: @project-version@\nCreated by Pigletoos of Skywall',
                        type = 'description'
                    },
                }
            }
        }
    }
end


UnderbellyBuddy.lua
local UnderbellyBuddy = LibStub('AceAddon-3.0'):NewAddon('UnderbellyBuddy', 'AceConsole-3.0')
local defaults = {
    profile = {
        enable = true,
        lock = false,
        warning = true,
        size = 1.5
    }
}

function UnderbellyBuddy:OnInitialize()
    self.db = LibStub('AceDB-3.0'):New('UnderbellyBuddyDB', defaults)  
    self.container = CreateFrame('Frame', 'UnderbellyBuddyTimerBar', UIParent)
    self.container:SetSize(self.db.profile.size * 140, self.db.profile.size * 20)
    self.container:SetMovable(true)
    self.container:SetUserPlaced(true)
    self.container:SetPoint('CENTER', 0, 150)
    self.container:EnableMouse(true)
    self.container:RegisterForDrag('LeftButton')
    self.container:SetScript('OnDragStart', function(self) if not UnderbellyBuddy.db.profile.lock then self:StartMoving() end end)
    self.container:SetScript('OnDragStop', function(self) self:StopMovingOrSizing() end)
    self.container:SetScript('OnMouseDown', function(self, button) if button == 'RightButton' then self:Hide() UnderbellyBuddy.bar.timer:Hide() end end)
end


Modules/HiredGuard.lua
local UnderbellyBuddy = LibStub('AceAddon-3.0'):GetAddon('UnderbellyBuddy')
local UnderbellyBuddyHiredGuard = UnderbellyBuddy:NewModule('UnderbellyBuddyHiredGuard', 'AceConsole-3.0', 'AceEvent-3.0')
local buffName, _, buffIcon = GetSpellInfo(203894)
local fontName = GameFontHighlightSmallOutline:GetFont()
local inUnderbelly = false
local hasGuardBuff = false
local barSize = {}
local secondsToDisplayWarning = {30, 90}
local barBase = {
    width = 140,
    height = 20,
    font = 8
}
local run = {
    warning = {},
    bar = false
}

function UnderbellyBuddyHiredGuard:OnInitialize()
    self.db = UnderbellyBuddy.db

    barSize = {
        width = self.db.profile.size * barBase.width,
        height = self.db.profile.size * barBase.height,
        font = self.db.profile.size * barBase.font
    }

    for _, value in pairs(secondsToDisplayWarning) do
        run.warning[value] = false
    end

    self:RegisterChatCommand('ubshow', 'ShowBar')
    self:RegisterChatCommand('ubhide', 'HideBar')
end

function UnderbellyBuddyHiredGuard:OnEnable()
    local candyBar = LibStub('LibCandyBar-3.0')

    UnderbellyBuddy.bar = {}

    if not UnderbellyBuddy.bar.timer then
        UnderbellyBuddy.bar.timer = candyBar:New('Interface\\AddOns\\UnderbellyBuddy\\Media\\bar', barSize.width, barSize.height)
        UnderbellyBuddy.bar.timer:SetPoint('CENTER', UnderbellyBuddy.container)
        UnderbellyBuddy.bar.timer:SetLabel(buffName)
        UnderbellyBuddy.bar.timer:SetIcon(buffIcon)
        UnderbellyBuddy.bar.timer.candyBarLabel:SetFont(fontName, barSize.font)
        UnderbellyBuddy.bar.timer.candyBarDuration:SetFont(fontName, barSize.font)
        UnderbellyBuddy.bar.timer:Hide()
    end

    if not UnderbellyBuddy.bar.test then
        UnderbellyBuddy.bar.test = candyBar:New('Interface\\AddOns\\UnderbellyBuddy\\Media\\bar', barSize.width, barSize.height)
        UnderbellyBuddy.bar.test:SetPoint('CENTER', UnderbellyBuddy.container)
        UnderbellyBuddy.bar.test:SetLabel('Test Bar')
        UnderbellyBuddy.bar.test:SetIcon(buffIcon)
        UnderbellyBuddy.bar.test.candyBarLabel:SetFont(fontName, barSize.font)
        UnderbellyBuddy.bar.test.candyBarDuration:SetFont(fontName, barSize.font)
        UnderbellyBuddy.bar.test:Hide()
    end

    self:RegisterEvent('PLAYER_ENTERING_WORLD', 'CheckBodyGuard')
    self:RegisterEvent('ZONE_CHANGED_NEW_AREA', 'CheckBodyGuard')
    self:RegisterEvent('UNIT_AURA')
end

function UnderbellyBuddyHiredGuard:OnDisable()
    self:UnregisterEvent('PLAYER_ENTERING_WORLD')
    self:UnregisterEvent('ZONE_CHANGED_NEW_AREA')
    self:UnregisterEvent('UNIT_AURA')
    self:StopBar()
end

function UnderbellyBuddyHiredGuard:CheckBodyGuard()
    inUnderbelly = self:CheckZone(GetSubZoneText())
    hasGuardBuff = self:CheckBuff()
end

function UnderbellyBuddyHiredGuard:UNIT_AURA(eventName, unit)
    self:StartBar()
end

function UnderbellyBuddyHiredGuard:CheckZone(subzone)
    local correctZones = {'The Underbelly', 'The Underbelly Descent', 'Circle of Wills', 'The Black Market'}
    local value

    for _, value in pairs(correctZones) do
        if subzone == value then
            return true
        end
    end

    return false
end

function UnderbellyBuddyHiredGuard:CheckBuff()
    local buff = UnitBuff('player', buffName)

    if buff == buffName then
        return true
    else
        return false
    end
end

function UnderbellyBuddyHiredGuard:StartBar()
    if inUnderbelly then
        hasGuardBuff = self:CheckBuff()

        if hasGuardBuff and not run.bar then
            UnderbellyBuddy.bar.test:Hide()
            UnderbellyBuddy.bar.timer:AddUpdateFunction(function(bar) self:CheckRemaingTime(bar) end)
            UnderbellyBuddy.bar.timer:SetDuration(300)
            UnderbellyBuddy.bar.timer:Start()
        end

        if not hasGuardBuff then
            self:StopBar()
        end
    else
        self:StopBar()
    end
end

function UnderbellyBuddyHiredGuard:StopBar()
    if run.bar then
        UnderbellyBuddy.bar.timer:Stop()
        run.bar = false

        for _, value in pairs(secondsToDisplayWarning) do
            run.warning[value] = false
        end
    end
end

function UnderbellyBuddyHiredGuard:Round(num, numDecimalPlaces)
    local mult = 10^(numDecimalPlaces or 0)
    return math.floor(num * mult + 0.5) / mult
end

function UnderbellyBuddyHiredGuard:CheckRemaingTime(bar)
    if self.db.profile.warning then
        local timeLeft = self:Round(bar.remaining, 1)
        local timeDisplay = string.format('%d seconds', timeLeft)

        if timeLeft > 60 then
            timeDisplay = string.format('%d minute %d seconds', timeLeft / 60 % 60, timeLeft % 60)         
        end

        for _, value in pairs(secondsToDisplayWarning) do
            if timeLeft == value and not run.warning[value] then
                PlaySoundFile('Sound\\Interface\\RaidWarning.ogg') 
                RaidNotice_AddMessage(RaidWarningFrame, string.format('Only %s remaining for %s!', timeDisplay, buffName), ChatTypeInfo['RAID_WARNING'])
                run.warning[value] = true
            end
        end
    end

    run.bar = true
end

function UnderbellyBuddyHiredGuard:SetEnabled(_, value)
    self.db.profile.enable = value
    
    if self.db.profile.enable then
        self:OnEnable()
    else
        self:OnDisable()
    end
end

function UnderbellyBuddyHiredGuard:ShowBar() 
    if run.bar then
        UnderbellyBuddy.container:Show()  
        UnderbellyBuddy.bar.timer:Show() 
    else 
        UnderbellyBuddy:Print('No bar to display') 
    end 
end

function UnderbellyBuddyHiredGuard:HideBar()
    if run.bar then
        UnderbellyBuddy.container:Hide() 
        UnderbellyBuddy.bar.timer:Hide() 
    else 
        UnderbellyBuddy:Print('No bar to hide') 
    end 
 end

function UnderbellyBuddyHiredGuard:ShowTestBar()
    if not run.bar then
        UnderbellyBuddy.bar.test:SetDuration(20)
        UnderbellyBuddy.bar.test:Start()
    else
        UnderbellyBuddy:Print(string.format('%s is in progress and the test bar cannot be displayed.', buffName))
    end
end

function UnderbellyBuddyHiredGuard:SetSize(_, value) 
    self.db.profile.size = value
    
    UnderbellyBuddy.bar.timer:SetSize(value * barBase.width, value * barBase.height)
    UnderbellyBuddy.bar.timer.candyBarLabel:SetFont(fontName, value * barBase.font)
    UnderbellyBuddy.bar.timer.candyBarDuration:SetFont(fontName, value * barBase.font)

    UnderbellyBuddy.bar.test:SetSize(value * barBase.width, value * barBase.height)
    UnderbellyBuddy.bar.test.candyBarLabel:SetFont(fontName, value * barBase.font)
    UnderbellyBuddy.bar.test.candyBarDuration:SetFont(fontName, value * barBase.font)
end
